lib/client.ts

```diff
@@ -17,58 +24,156 @@ export class Client {
     const options: {
-      state: string;
       body?: {
         code_verifier: string;
       };
-    } = {
-      state: this.state
-    };
+    } = {};
 
-    if (opts.codeVerifier) {
+    if (opts.verifier) {
       options.body = {
-        code_verifier: opts.codeVerifier
+        code_verifier: opts.verifier
       };
     }
 
-    const tokenData = await this.oauth2Client.code.getToken(
-      callbackUri,
-      options
+    const token = await this.oauth2Client.code.getToken(callbackUri, options);
+
+    this.token = token;
+
+    return token;
+  };
+
+  /**
+   * Use a previously created token for all upcoming requests.
+   */
+  public setToken = (
+    accessToken: string,
+    refreshToken?: string,
+    tokenType?: string
+  ): ClientOAuth2.Token => {
+    const data = {};
+    let token;
+
+    if (tokenType && refreshToken) {
+      token = this.oauth2Client.createToken(
+        accessToken,
+        refreshToken,
+        tokenType,
+        data
+      );
+    } else if (refreshToken) {
+      token = this.oauth2Client.createToken(accessToken, refreshToken, data);
+    } else {
+      token = this.oauth2Client.createToken(accessToken, data);
+    }
+
+    this.token = token;
+    return token;
+  };
+
+  /**
+   * Send a raw GraphQL request and return its response.
+   */
+  public rawQuery = async (
+    query: string,
+    variables?: Variables
+  ): Promise<Query> => {
+    if (!this.token) {
+      throw new Error("User unauthorized");
+    }
+
+    this.graphQLClient.setHeader(
+      "Authorization",
+      `Bearer ${this.token.accessToken}`
     );
-    return tokenData;
+
+    const { data } = await this.graphQLClient.rawRequest(query, variables);
+
+    return data;
+  };
+
+  public getAllTransactions = async (): Promise<Array<TransactionListItem>> => {
+    const query = `{
+      viewer {
+          mainAccount {
+            transactions {
+              edges {
+                node {
+                  id
+                  amount
+                  name
+                  iban
+                  type
+                  bookingDate
+                  valutaDate
+                  originalAmount
+                  foreignCurrency
+                  e2eId
+                  mandateNumber
+                  paymentMethod
+                  category
+                  userSelectedBookingDate
+                  purpose
+                  bookingType
+                  invoiceNumber
+                  invoicePreviewUrl
+                  invoiceDownloadUrl
+                  documentType
+                }
+              }
+            }
+          }
+        }
+      }`;
+
+    const result: Query = await this.rawQuery(query);
+
+    const transactions = (
+      (result &&
+        result.viewer &&
+        result.viewer.mainAccount &&
+        result.viewer.mainAccount.transactions &&
+        result.viewer.mainAccount.transactions.edges) ||
+      []
+    ).map(node => node.node);
```

Comment:
Looks a bit ugly, can we e.g. use lodash until we can use TypeScript 3.7 and the nullish coalescing operator `??`? so it might look like `_.get(result, "viewer.mainAccount.transactions.edges")`

Reply:
sure, done in https://github.com/kontist/sdk/pull/2/commits/e9013b343956d1cafd39dfcf0df4ecad6a580603